.github/workflows/test.yml

```
@@ -12,8 +18,27 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
+      - name: Get permission
+        id: checkAccess
+        uses: actions-cool/check-user-permission@7b90a27f92f3961b368376107661682c441f6103 # v2.3.0
+        with:
+          require: write
+          username: ${{ github.triggering_actor }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Check permission
+        if: steps.checkAccess.outputs.require-result == 'false' && github.actor != 'dependabot[bot]'
+        run: |
+          echo "${{ github.triggering_actor }} does not have permissions on this repo."
+          echo "Current permission level is ${{ steps.checkAccess.outputs.user-permission }}"
+          echo "Job originally triggered by ${{ github.actor }}"
+          exit 1
+
       - name: Checkout
         uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
```

I was a little curious about this particular change... Typically in a PR you would want to checkout the merge commit, rather than the head. Using the pr head will test only what is on the head branch, and will not incorporate commits that have already been merged to the base branch. Considering the other risk mitigations here, using `pull_request_target` and testing the actor permissions, do you mean to use the merge commit?

```suggestion
          ref: ${{ github.event.pull_request.merge_commit_sha }}
```